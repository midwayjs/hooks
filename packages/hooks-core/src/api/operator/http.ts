import { BaseTrigger, MetadataHelper, Operator, OperatorType } from '../type'

export const HttpTriggerType = 'HTTP'

export enum HttpMetadata {
  METHOD = 'Http_Method',
  QUERY = 'Http_Query',
  PARAMS = 'Http_Params',
  HEADERS = 'Http_Headers',
  RESPONSE = 'Http_Response',
}

export enum ResponseMetaType {
  CODE = 'Http_Response_Code',
  HEADER = 'Http_Response_Header',
  CONTENT_TYPE = 'Http_Response_ContentType',
  REDIRECT = 'Http_Response_Redirect',
}

export enum HttpMethod {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  DELETE = 'DELETE',
  PATCH = 'PATCH',
  HEAD = 'HEAD',
  OPTIONS = 'OPTIONS',
  ALL = 'ALL',
}

export interface HttpTrigger extends BaseTrigger {
  type: typeof HttpTriggerType
  method: HttpMethod
  path?: string
}

export type HttpInputMetadata = {
  query?: Record<string, string>
  headers?: Record<string, string>
  params?: Record<string, string>
}

function createHttpMethodOperator(method: HttpMethod) {
  return (path?: string) => {
    return {
      name: method,
      metadata({ setMetadata }) {
        setMetadata<HttpTrigger>(OperatorType.Trigger, {
          type: HttpTriggerType,
          method,
          path,
          requestClient: {
            fetcher: 'http',
            client: '@midwayjs/rpc',
          },
        })
      },
    } as Operator<void>
  }
}

// HTTP Method
export const All = createHttpMethodOperator(HttpMethod.ALL)
export const Get = createHttpMethodOperator(HttpMethod.GET)
export const Post = createHttpMethodOperator(HttpMethod.POST)
export const Delete = createHttpMethodOperator(HttpMethod.DELETE)
export const Put = createHttpMethodOperator(HttpMethod.PUT)
export const Patch = createHttpMethodOperator(HttpMethod.PATCH)
export const Head = createHttpMethodOperator(HttpMethod.HEAD)
export const Options = createHttpMethodOperator(HttpMethod.OPTIONS)

// HTTP Helper
export function Query<T extends Record<string, string>>(): Operator<{
  query: T
}> {
  return {
    name: HttpMetadata.QUERY,
    input: true,
    metadata({ setMetadata }) {
      setMetadata(HttpMetadata.QUERY, true)
    },
  }
}

export function Params<T extends Record<string, string>>(): Operator<{
  params: T
}> {
  return {
    name: HttpMetadata.PARAMS,
    input: true,
    metadata({ setMetadata }) {
      setMetadata(HttpMetadata.PARAMS, true)
    },
  }
}

export function Headers<T extends Record<string, string>>(): Operator<{
  headers: T
}> {
  return {
    name: HttpMetadata.HEADERS,
    input: true,
    metadata({ setMetadata }) {
      setMetadata(HttpMetadata.HEADERS, true)
    },
  }
}

export type ResponseMetaData = {
  type: ResponseMetaType
  code?: number
  header?: {
    key: string
    value: string
  }
  url?: string
  contentType?: string
}

export function HttpCode(code: number): Operator<void> {
  return {
    name: 'HttpCode',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetaType.CODE, { code })
    },
  }
}

export function SetHeader(key: string, value: string): Operator<void> {
  return {
    name: 'SetHeader',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetaType.HEADER, {
        header: {
          key,
          value,
        },
      })
    },
  }
}

export function Redirect(url: string, code?: number): Operator<void> {
  return {
    name: 'Redirect',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetaType.REDIRECT, {
        url,
        code,
      })
    },
  }
}

export function ContentType(contentType: string): Operator<void> {
  return {
    name: 'ContentType',
    metadata(helper) {
      setResponseMetaData(helper, ResponseMetaType.CONTENT_TYPE, {
        contentType,
      })
    },
  }
}

function setResponseMetaData(
  helper: MetadataHelper,
  type: ResponseMetaType,
  value: Partial<ResponseMetaData>
) {
  const responseMetaData =
    helper.getMetadata<ResponseMetaData[]>(HttpMetadata.RESPONSE) || []

  helper.setMetadata(HttpMetadata.RESPONSE, [
    ...responseMetaData,
    {
      type,
      ...value,
    },
  ])
}
