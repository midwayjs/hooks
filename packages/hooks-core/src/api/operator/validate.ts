import { z } from 'zod'
import { validateFunction } from '../../'
import { Operator } from '../type'

/**
 * Validate input arguments.
 * Default lib is zod.
 */
export type Validator = (schemas: any[], input: any[]) => any | Promise<any>

let validator: Validator = (schemas: any, input: any[]) => {
  const tuple = z.tuple(schemas as any)
  return tuple.parseAsync(input)
}

export function setValidator(newValidator: Validator) {
  validateFunction(newValidator, 'newValidator')
  validator = newValidator
}

export function Validate(...schemas: any[]): Operator<void> {
  return {
    name: 'Validate',
    async execute({ getInputArguments }, next) {
      validateFunction(validator, 'validator')
      const inputs = getInputArguments()
      await validator(schemas, inputs)
      return next()
    },
  }
}
