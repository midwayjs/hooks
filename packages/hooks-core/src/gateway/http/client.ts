import { init, parse } from 'es-module-lexer'

import { ApiParam, formatCode } from '../../index'
import { HTTPRouter } from './router'

interface RenderParam extends Partial<ApiParam> {
  isExportDefault?: boolean
  name?: string
}

export async function createHTTPApiClient(
  file: string,
  code: string,
  router: HTTPRouter
) {
  await init
  const [, exports] = parse(code)

  if (exports.length === 0) {
    return null
  }

  const baseUrl = router.getBaseUrl(file)
  const {
    request: { client },
  } = router.projectConfig

  const funcs: RenderParam[] = exports.map((name) => {
    const isExportDefault = name === 'default'
    return {
      url: getUrl(baseUrl, name),
      meta: {},
      functionId: router.getFunctionId(file, name, isExportDefault),
      isExportDefault: isExportDefault,
      name: isExportDefault ? '$default' : name,
    }
  })

  return formatCode(buildClient(client, funcs))
}

function buildClient(client: string, funcs: RenderParam[]) {
  const buildFunction = (func: RenderParam) => {
    return `
      export ${func.isExportDefault ? 'default' : ''} function ${
      func.name
    } (...args) {
        return request({
          url: '${func.url}',
          method: args.length === 0 ? 'GET' : 'POST',
          data: { args },
          meta: ${JSON.stringify(func.meta, null, 2)}
        })
      }
    `
  }

  const functions = funcs.map(buildFunction).join('\n')

  return `
    import { request } from '${client}';
    ${functions}
  `
}

function getUrl(baseUrl: string, name: string) {
  if (name === 'default') {
    return baseUrl
  }

  if (baseUrl.endsWith('/')) {
    return `${baseUrl}${name}`
  }

  return `${baseUrl}/${name}`
}
