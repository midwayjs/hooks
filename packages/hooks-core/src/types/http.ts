/**
 * Now the function accepts all http methods
 * @deprecated will be removed in next major version.
 */
export type ApiHttpMethod = 'GET' | 'POST'

export type ApiBaseParam = {
  /**
   * @description generated by the file router, base on file path
   */
  functionId?: string
  data?: {
    args: any[]
    [key: string]: any
  }
}

export interface ApiParam extends ApiBaseParam {
  url: string
  /**
   * Now the function accepts all http methods
   * @deprecated will be removed in next major version.
   */
  method: ApiHttpMethod

  meta: {
    functionName?: string
    [key: string]: any
  }
}
