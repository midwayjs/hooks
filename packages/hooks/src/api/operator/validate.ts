import { z } from 'zod'
import { HttpStatus, MidwayHttpError, registerErrorCode } from '@midwayjs/core'
import { Operator, setValidator, useContext } from '@midwayjs/hooks-core'

const HooksValidateErrorCode = registerErrorCode('Validation', {
  FAILED: 'FAILED',
})

class HooksValidationError extends MidwayHttpError {
  constructor(message: string, status: number, cause: any) {
    super(message, status, HooksValidateErrorCode.FAILED, { cause })
  }
}

export { Validate } from '@midwayjs/hooks-core'

setValidator(async (schemas: any, inputs: any[]) => {
  const result = await z.tuple(schemas).safeParseAsync(inputs)

  if (result.success === false) {
    throw new HooksValidationError(
      result.error.message,
      HttpStatus.UNPROCESSABLE_ENTITY,
      result.error
    )
  }
})

type ValidateHttpOption = {
  query?: z.Schema<any>
  data?: z.Schema<any>[]
  params?: z.Schema<any>
  headers?: z.Schema<any>
}

export function ValidateHttp(options: ValidateHttpOption): Operator<void> {
  return {
    name: 'ValidateHttp',
    async execute({ getInputArguments }, next) {
      const ctx = useContext()

      try {
        if (options.params) await options.params.parseAsync(ctx.params)
        if (options.query) await options.query.parseAsync(ctx.query)
        if (options.headers) await options.headers.parseAsync(ctx.headers)
        if (options.data) {
          const inputs = getInputArguments()
          const tuple = z.tuple(options.data as any)
          await tuple.parseAsync(inputs)
        }
      } catch (error) {
        throw new HooksValidationError(
          error.message,
          HttpStatus.UNPROCESSABLE_ENTITY,
          error
        )
      }

      return next()
    },
  }
}
