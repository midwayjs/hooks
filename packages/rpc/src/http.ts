import type { FileRecord, HttpRequestOptions } from '@midwayjs/hooks-core'
import { args } from './util'
import { createClient, RequestOptionsCreator } from './client'
import { HttpContext, Middleware, SetupOptions } from './type'
import axios, { Options } from 'redaxios'
import fetch from 'isomorphic-unfetch'

export const client: SetupOptions = {
  baseURL: '',
  async fetcher(req: HttpRequestOptions, options: SetupOptions) {
    const res = await axios({
      method: req.method as Options['method'],
      url: req.url,
      data: req.data,
      params: req.query,
      headers: req.headers,
      baseURL: options.baseURL,
      withCredentials: options.withCredentials,
      fetch,
    })
    return res.data
  },
  middleware: [],
  withCredentials: false,
}

export function setupHttpClient(options: SetupOptions) {
  Object.assign(client, options)
}

export const creator: RequestOptionsCreator<HttpRequestOptions> = (req) => {
  const { trigger, metadata, args: inputs } = req

  return {
    method: trigger.method,
    url:
      typeof metadata?.params === 'object'
        ? format(trigger.path, metadata.params)
        : trigger.path,
    data: inputs.length > 0 ? args(...inputs) : null,

    query: metadata?.query,
    headers: {
      accept: 'application/json',
      ...metadata?.headers,
    },
    files: metadata?.files,
  }
}

const request: Middleware<HttpContext> = async (ctx, next) => {
  ctx.res = await client.fetcher(ctx.req, client)
  return next()
}

const formdata: Middleware<HttpContext> = async (ctx, next) => {
  const files: FileRecord = ctx.req.files
  if (!files) {
    throw new Error('no files')
  }

  const formdata = new FormData()
  for (const [key, value] of Object.entries(files)) {
    if (value instanceof FileList) {
      for (const file of value) {
        formdata.append(key, file)
      }
    } else {
      formdata.append(key, value)
    }
  }

  ctx.req.data = formdata

  return next()
}

export const http = createClient<HttpContext>(creator, () => [
  ...client.middleware,
  request,
])

export const upload = createClient<HttpContext>(creator, () => [
  ...client.middleware,
  formdata,
  request,
])

export function format(url: string, params: Record<string, string>) {
  let result = url
  for (const key in params) {
    result = result.replace(`:${key}`, params[key])
  }
  return result
}
