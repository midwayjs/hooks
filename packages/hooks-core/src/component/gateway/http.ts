import { existsSync } from 'fs'
import staticCache from 'koa-static-cache'
import {
  HTTPRoute,
  ServerRoute,
} from 'packages/hooks-core/dist/midwayjs-hooks-core.cjs'
import { __decorate } from 'tslib'
import { join } from 'upath'

import {
  IMidwayApplication,
  IMidwayContainer,
  IMidwayContext,
} from '@midwayjs/core'
import { All, Controller, Inject, Provide } from '@midwayjs/decorator'

import { superjson } from '../../lib'
import { useContext } from '../../runtime'
import { ApiFunction } from '../../types/common'
import {
  ComponentConfig,
  CreateApiParam,
  HooksGatewayAdapter,
} from '../../types/gateway'
import { isDevelopment, useHooksMiddleware } from '../../util'

export class HTTPGateway implements HooksGatewayAdapter {
  config: ComponentConfig
  container: IMidwayContainer
  app: IMidwayApplication<IMidwayContext>

  constructor(config: ComponentConfig) {
    this.config = config
  }

  is(route: ServerRoute<HTTPRoute>) {
    return !!route?.basePath
  }

  createApi(param: CreateApiParam) {
    const { id, fn, httpPath } = param

    // 设置中间件
    fn.middleware.unshift(...this.getMiddleware())

    // Source: https://www.typescriptlang.org/play?noImplicitAny=false&strictNullChecks=false&strictFunctionTypes=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&noImplicitReturns=false&alwaysStrict=false&importHelpers=true&emitDecoratorMetadata=false&ts=4.1.5#code/JYWwDg9gTgLgBAbzgSQHYCsCmBjGAaOAYQlRiggBsLMoCBBKggBXIDdgATTOAXzgDNyIOAHIAAiE4B3AIYBPdAGcA9F2zQZMaCIBQO9akXx+qOAF44ACgB0tmVADmigFxwZqOQG0AugEpzAHyIPHpiLBDsXJa+OmLEpORUNJYiyiIx2BQyiopwAGIArqi4wCTxMDLAqDSIOnD1cGJoWLjRdQ24AB6u7nJ6DY0MFCkiBEiSHBzUslCYrj68MQPZcsVwABbuUzXRtQP11PD2TuZwMOvAitZd1rMAjgWYRtYARhAcctbHuQA+P3A+dr7OCzGAFKCmGSyYDGVA2OyORRLBohEJAA
    let FunctionContainer = class FunctionContainer {
      ctx: any
      async handler() {
        let args = this.ctx.request?.body?.args || []
        return await fn(...args)
      }
    }
    __decorate([Inject()], FunctionContainer.prototype, 'ctx', void 0)
    __decorate(
      [All(httpPath, { middleware: fn.middleware })],
      FunctionContainer.prototype,
      'handler',
      null
    )
    FunctionContainer = __decorate(
      [Provide(id), Controller('/')],
      FunctionContainer
    )

    this.container.bind(id, FunctionContainer)
  }

  getMiddleware() {
    const mws = [this.handleError]

    const enableSuperjson = this.config.internal.superjson
    if (enableSuperjson) {
      mws.push(this.deserializeSuperjson)
    }

    return mws.map(useHooksMiddleware)
  }

  async deserializeSuperjson(next: any) {
    await next()
    const ctx = useContext()
    if (ctx.type.includes('application/json')) {
      ctx.body = superjson.serialize(ctx.body)
    }
  }

  handleError = async (next: any) => {
    try {
      await next()
    } catch (error) {
      const ctx = useContext()

      if (this.config.internal.superjson) {
        ctx.status = 500
        ctx.body = superjson.serialize(error)
      } else {
        throw error
      }
    }
  }

  afterCreate() {
    if (isDevelopment()) {
      return
    }

    if (!this.isViteProject) {
      return
    }

    const {
      router: { routes },
    } = this.config
    if (routes.has('/') || routes.has('/*')) {
      return
    }

    const baseDir = this.container.get('baseDir')
    const mw = staticCache({
      dir: join(baseDir, '..', this.config.internal.build.viteOutDir),
      dynamic: true,
      alias: {
        '/': 'index.html',
        /**
         * Add alias for windows, '/' -> '\\'
         * https://github.com/koajs/static-cache/blob/master/index.js#L45
         */
        '\\': 'index.html',
      },
      buffer: true,
      gzip: true,
    })

    const fn: ApiFunction = async () => {}
    fn.middleware = [mw]

    this.createApi({
      id: 'hooks-host',
      httpPath: '/*',
      fn,
    })
  }

  get isViteProject() {
    return ['vite.config.ts', 'vite.config.js'].some((config) =>
      existsSync(join(this.config.root, config))
    )
  }
}
