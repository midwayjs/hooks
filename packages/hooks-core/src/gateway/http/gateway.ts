import { existsSync } from 'fs'
import staticCache from 'koa-static-cache'
import { __decorate } from 'tslib'
import { join } from 'upath'

import { IMidwayContainer } from '@midwayjs/core'
import { All, Controller } from '@midwayjs/decorator'

import { createFunctionContainer, useContext } from '../../runtime'
import { ApiFunction } from '../../types/common'
import { Route } from '../../types/config'
import {
  ComponentOptions,
  CreateApiOptions,
  HooksGatewayAdapter,
} from '../../types/gateway'
import { isDevelopment, useHooksMiddleware } from '../../util'
import { createHTTPApiClient } from './client'
import { HTTPRouter } from './router'

export class HTTPGateway implements HooksGatewayAdapter {
  static is(route: Route) {
    return !!route?.basePath
  }

  static router = HTTPRouter
  static createApiClient = createHTTPApiClient

  options: ComponentOptions
  container: IMidwayContainer
  router: HTTPRouter

  constructor(options: ComponentOptions) {
    this.options = options
    this.router = new HTTPRouter({
      root: this.options.root,
      projectConfig: this.options.projectConfig,
      useSourceFile: this.options.router.useSourceFile,
    })
  }

  createApi(options: CreateApiOptions) {
    const { file, functionName, isExportDefault, fn } = options
    const httpPath = this.router.getHTTPPath(
      file,
      functionName,
      isExportDefault
    )

    fn._param.url = httpPath
    this.createHTTPApi(httpPath, options)
  }

  createHTTPApi(httpPath: string, options: CreateApiOptions) {
    const { functionId, fn } = options
    // setup middleware
    const middleware = [...fn.middleware]

    const FunctionContainer = createFunctionContainer({
      isHTTP: true,
      fn,
      functionId,
      parseArgs(ctx) {
        const args = ctx.request?.body?.args || []
        return args
      },
      classDecorators: [Controller('/')],
      handlerDecorators: [All(httpPath, { middleware })],
    })

    this.container.bind(functionId, FunctionContainer)
  }

  afterCreate() {
    if (isDevelopment()) {
      return
    }

    if (!this.isViteProject) {
      return
    }

    const { routes } = this.router
    if (routes.has('/') || routes.has('/*')) {
      return
    }

    const baseDir = this.container.get('baseDir')
    const mw = staticCache({
      dir: join(baseDir, '..', this.options.projectConfig.build.viteOutDir),
      dynamic: true,
      alias: {
        '/': 'index.html',
        /**
         * Add alias for windows, '/' -> '\\'
         * https://github.com/koajs/static-cache/blob/master/index.js#L45
         */
        '\\': 'index.html',
      },
      buffer: true,
      gzip: true,
    })

    const fn: ApiFunction = async () => {}
    fn.middleware = [mw]

    this.createHTTPApi('/*', {
      functionId: 'hooks-host',
      fn,
    })
  }

  get isViteProject() {
    return ['vite.config.ts', 'vite.config.js'].some((config) =>
      existsSync(join(this.options.root, config))
    )
  }
}
