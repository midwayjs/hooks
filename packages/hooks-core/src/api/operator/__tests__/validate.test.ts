import { z } from 'zod'
import { Api } from '../..'
import { Post } from '../http'
import { setValidator, Validate, Validator } from '../validate'

describe('validator', () => {
  it('use default validator', async () => {
    const schema = z.number()

    const fn = Api(
      Post(),
      Validate(schema),
      async (count: z.infer<typeof schema>) => {
        return count
      }
    )

    expect(await fn(1)).toEqual(1)
  })

  it('should throw error', async () => {
    const schema = z.object({
      name: z.string(),
      age: z.number(),
    })

    const fn = Api(
      Post(),
      Validate(schema),
      async (user: z.infer<typeof schema>) => {
        return user
      }
    )

    expect(fn({ name: 'hooks', age: 1 })).toBeTruthy()
    expect(
      fn({ name: 1, age: 1 } as any)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(
      fn({ name: '', age: '' } as any)
    ).rejects.toThrowErrorMatchingSnapshot()
  })

  it('should support custom validator', async () => {
    const validator: Validator = (schemas, inputs) => {
      if (schemas[0] === String) {
        if (typeof inputs[0] !== 'string') {
          throw new Error('input must be a string')
        }
      } else {
        throw new Error('unsupported schema')
      }
    }

    setValidator(validator)

    const fn = Api(Post(), Validate(String), async (name: string) => {
      return { name }
    })

    expect(await fn('hooks')).toEqual({ name: 'hooks' })
    expect(fn(1 as any)).rejects.toThrowErrorMatchingSnapshot()

    const fn2 = Api(Post(), Validate(Number), async (age: number) => {
      return { age }
    })
    expect(fn2(1)).rejects.toThrowErrorMatchingSnapshot()
  })
})
