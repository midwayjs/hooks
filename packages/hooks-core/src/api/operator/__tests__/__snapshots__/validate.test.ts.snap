// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validator should support custom validator 1`] = `"input must be a string"`;

exports[`validator should support custom validator 2`] = `"unsupported schema"`;

exports[`validator should throw error 1`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"string\\",
    \\"received\\": \\"number\\",
    \\"path\\": [
      0,
      \\"name\\"
    ],
    \\"message\\": \\"Expected string, received number\\"
  }
]"
`;

exports[`validator should throw error 2`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"number\\",
    \\"received\\": \\"string\\",
    \\"path\\": [
      0,
      \\"age\\"
    ],
    \\"message\\": \\"Expected number, received string\\"
  }
]"
`;
